[quant] Create default custom modules for LSTM and MHA

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #71192
* #71191
* #71190
* #71189



Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

HDCharles committed:
[quant] Create default custom modules for LSTM and MHA

Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"


Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"


Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"


Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"


Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!


Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33452890).

z-a-f commented on torch/ao/quantization/quantize.py:
```diff
@@ -261,7 +277,7 @@ def prepare(model, inplace=False, allow_list=None,
     """
     torch._C._log_api_usage_once("quantization_api.quantize.prepare")
     if prepare_custom_config_dict is None:
-        prepare_custom_config_dict = {}
+        prepare_custom_config_dict = get_default_custom_config_dict()
```
I think it's better to keep the defaults as well. Something like this:

```python
_prepare_custom_config_dict = get_default_custom_config_dict()
_prepare_custom_config_dict.update(prepare_custom_config_dict or {})
prepare_custom_config_dict = _prepare_custom_config_dict

for key, value in prepare_custom_config_dict['float_to_observed_custom_module_class').items():
    if value is None:
         prepare_custom_config_dict['float_to_observed_custom_module_class').pop(key)

for key, value in prepare_custom_config_dict['observed_to_quantized_custom_module_class').items():
    if value is None:
         prepare_custom_config_dict['observed_to_quantized_custom_module_class').pop(key)
```

(Of course you can rewrite it to avoid the temporary `_prepare_custom_config_dict`).


The benefit of it is that if the user specifies custom config dict that does not involve the LSTM or MHA, we still want to support them:

```python
custom_module_config = {
    'float_to_observed_custom_module_class': {
        MyModule: MyOtherModule
    },
    'observed_to_quantized_custom_module_class': {
        MyOtherModule: MyOtherOtherModule
    }
}
prepared = torch.ao.quantization.prepare(model, prepare_custom_config_dict=custom_module_config)
```

In the above example, if the `model` has an instance of an `nn.LSTM`, we still want to be able to replace it.

The downside of using it by default is that if the user doesn't want to quantize the LSTM, they would need to be explicit about it:

```python
custom_module_config = {
    'float_to_observed_custom_module_class': {
        torch.nn.LSTM: None
    },
    'observed_to_quantized_custom_module_class': {
        torch.nn.quantizable.LSTM: None
    }
}
prepared = torch.ao.quantization.prepare(model, prepare_custom_config_dict=custom_module_config)
```

What do you think?

HDCharles replied:
Per the proposal (https://docs.google.com/document/d/16upncHr8zLQaynosbskCZP522oTkEMWzioMHKp1x0_s/edit), one of the goals was to align with the other quantization apis i.e. user specified matching overwrites rather than updates the defaults.

z-a-f replied:
sgtm

z-a-f commented on torch/ao/quantization/quantize.py:
```diff
@@ -543,7 +559,7 @@ def _convert(
         mapping = get_default_static_quant_reference_module_mappings() if is_reference \
             else get_default_static_quant_module_mappings()
     if convert_custom_config_dict is None:
-        convert_custom_config_dict = {}
+        convert_custom_config_dict = get_default_custom_config_dict()
```
Same comment as above

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!


Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33452890).

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!


Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33452890).

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!


Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!


Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33452890).

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!


Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33452890).

z-a-f reviewed (approved):
Looks good to me

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!


Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33452890).

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!


Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33452890).

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!


Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33452890/)!


Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33452890).

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.



Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]

HDCharles committed:
Update on "[quant] Create default custom modules for LSTM and MHA"




Currently we expect the users to provide custom modules for LSTM and MHA. However, as we almost always ask the users to use those modules in the custom context, it is better to make this behavior default. In this case we try to align with the base quantization API, if the user specifies a custom_config_dict then that is used, however if the value is left as None then the default is used. If a user would like to both use the default and modify it, they have to do so manually, however the default is accessible by get_default_custom_config_dict

Additionally, the NS which uses prepare to insert custom observers for
its purposes had to be slightly modified to pass in an empty
custom_config_dict in order to avoid modifying the custom modules.

Differential Revision: [D33452890](https://our.internmc.facebook.com/intern/diff/D33452890)

[ghstack-poisoned]
